### PR TITLE
DMP-4352: Move duplicated event logic to event receipt

### PR DESCRIPTION
**Pull request by Ben-Edwards-cgi**
### Change description ###
# Summary of Git Diff

The modifications primarily introduce new functionalities for finding and managing duplicate events, while also refactoring existing components for better performance and maintainability.

### New Features
- **Duplicate Event Handling**: 
  - New methods `findDuplicateEventIds` added to `EventRepository` to identify duplicates based on `eventId` and `createdDateTime`.
  - Added tests in `EventRepositoryTest` to verify the functionality of finding duplicate event IDs.

- **Asynchronous Processing**: 
  - Introduced `AsyncEventProcessor` to handle event processing asynchronously, which includes the cleanup of current flag events and removal of duplicate events.

### Refactoring
- **EventDispatcher**:
  - Refactored to use `AsyncEventProcessor` instead of the `CleanupCurrentFlagEventProcessor`, allowing for asynchronous handling of events.

- **Remove Duplicate Events Processor**:
  - Modified the interface and implementation to process specific events rather than running a bulk cleanup method.
  - Adjusted the logic to delete duplicate events while retaining the most recent one.

### Deletions
- Removed the `RemoveDuplicatedEventsAutomatedTask` and associated configurations from the codebase.

### Test Cases
- Expanded test coverage for new functionalities, particularly around the processing of events and the handling of duplicates.
- Ensured that existing tests were updated to reflect the changes in the method signatures and processing logic.

### Database Migration
- Added a migration script to remove the `RemoveDuplicatedEvents` task from the database, ensuring that it no longer exists in the application’s task scheduler.

These updates enhance the system's ability to manage event data effectively, reduce redundancy, and improve performance through asynchronous processing.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/darts/event/service/impl/RemoveDuplicateEventsProcessorImpl.java [hunk 1]
```diff
@@ -1,55 +1,32 @@
...
-import static java.util.Comparator.comparing;
-import static java.util.stream.Collectors.groupingBy;
-import static java.util.stream.Collectors.toList;
-import static java.util.stream.StreamSupport.stream;
 
 @Service
 @Slf4j
 public class RemoveDuplicateEventsProcessorImpl implements RemoveDuplicateEventsProcessor {
 
     private static final int CHUNK_SIZE = 1000;
```
Comment (jackmaloney) [inline src/main/java/uk/gov/hmcts/darts/event/service/impl/RemoveDuplicateEventsProcessorImpl.java:19]: Don't need this

Comment (Ben-Edwards-cgi) [reply src/main/java/uk/gov/hmcts/darts/event/service/impl/RemoveDuplicateEventsProcessorImpl.java]: removed

File: src/main/java/uk/gov/hmcts/darts/event/service/impl/RemoveDuplicateEventsProcessorImpl.java [hunk 2]
```diff
@@ -58,54 +33,25 @@ public RemoveDuplicateEventsProcessorImpl(
 
     @Override
     @Transactional
-    public void processRemoveDuplicateEvents() {
-        if (clearUpWindow < 1) {
-            throw new DartsException(
-                "The duplicate event deletion task is not configured correctly. " +
-                    "The clear-up window must be at least 1 day but is configured as " + clearUpWindow);
+    public boolean processEvent(Integer eventId) {
```
Comment (jackmaloney) [inline src/main/java/uk/gov/hmcts/darts/event/service/impl/RemoveDuplicateEventsProcessorImpl.java:36]: Maybe this is abit subjective but is the method named appropriately here? Agreed, we are processing the event, but do you think it may be confusing to return `true` if a duplicate is found? Like the logic in `AsyncEventProcessor` only continues if this processEvent returns `false`, which seems wrong. 
Wonder if it should be renamed to something like `findAndRemoveDuplicateEvent` ? Or something along those lines.

File: src/main/java/uk/gov/hmcts/darts/event/service/impl/AsyncEventProcessor.java [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+@Service
+@AllArgsConstructor
+public class AsyncEventProcessor {
+    private final CleanupCurrentFlagEventProcessor cleanupCurrentFlagEventProcessor;
+    private final RemoveDuplicateEventsProcessorImpl removeDuplicateEventsProcessor;
+
+
+    @Async("eventTaskExecutor")
+    public void processEvent(Integer eventId) {
```
Comment (jackmaloney) [inline src/main/java/uk/gov/hmcts/darts/event/service/impl/AsyncEventProcessor.java:16]: Do we need an integration test to ensure both processes (duplicate check and event current cleanup) are carried out in a happy path scenario?

Comment (Ben-Edwards-cgi) [reply src/main/java/uk/gov/hmcts/darts/event/service/impl/AsyncEventProcessor.java:16]: Integration tests are probably excessive here but certainly unit tests for this method would add value will add some

**Review by jackmaloney — APPROVED** [review]

File: src/main/java/uk/gov/hmcts/darts/task/config/RemoveDuplicatedEventsAutomatedTaskConfig.java [hunk 4]
```diff
@@ -1,13 +0,0 @@
-package uk.gov.hmcts.darts.task.config;
-
-import lombok.Getter;
-import lombok.Setter;
-import org.springframework.boot.context.properties.ConfigurationProperties;
-import org.springframework.context.annotation.Configuration;
-
-@ConfigurationProperties("darts.automated.task.remove-duplicated-events")
```
Comment (chrisbellingham-hmcts) [inline src/main/java/uk/gov/hmcts/darts/task/config/RemoveDuplicatedEventsAutomatedTaskConfig.java:8]: Also eliminate the now redundant `remove-duplicated-events` config in `application.yaml`

**Review by chrisbellingham-hmcts — APPROVED** [review]

File: src/main/java/uk/gov/hmcts/darts/task/config/RemoveDuplicatedEventsAutomatedTaskConfig.java (hunk 4 above)
Comment (Ben-Edwards-cgi) [reply src/main/java/uk/gov/hmcts/darts/task/config/RemoveDuplicatedEventsAutomatedTaskConfig.java:8]: Good spot removed :)
